src/openMVG/multiview/solver_resection_p2pt_fabbri.cpp

```diff
@@ -0,0 +1,2741 @@
+	       [1] -> sigmas2_len[TS_MAX_LEN]
+
+	       `sigmasX_len` (single values)
+	             [0] = int, [1] = int, [2] = int, ...  */
+
+	T   (&sigmas1)[TS_MAX_LEN][TS_MAX_LEN] = (*out)[0];
+	T   (&sigmas2)[TS_MAX_LEN][TS_MAX_LEN] = (*out)[1];
+	unsigned (&sigmas1_len)[TS_MAX_LEN]         = (*out_len)[0];
+	T pose_out[10];
+	for (unsigned i = 0; i < ts_len; i++) {
+		sigmas1_len[i] = 0; 
+
+		fn_t(ts[i], pose_out);
+
+		//T &fvalue = pose_out[0]; // double-checked: not used in matlab
+		const T &A = pose_out[0], &B = pose_out[1], &C = pose_out[2], 
+            &E = pose_out[3], &F = pose_out[4], &G = pose_out[5], &H = pose_out[6],
+            &J = pose_out[7], &K = pose_out[8], &L = pose_out[9];
+
+		std::complex<T> delta = sqrt(B*B - 4*A*C);
+		std::complex<T> sigma1_m = (-B - delta)/(2*A);
+		std::complex<T> sigma1_p = (-B + delta)/(2*A);
+
+		delta = sqrt(F*F - 4*E*G);
+		std::complex<T> sigma2_m = (-F - delta)/(2*E);
+		std::complex<T> sigma2_p = (-F + delta)/(2*E);
+
+		//% handle case of negative delta
+		if (std::abs(std::imag(sigma1_m)) < 1e-4) {
+			sigma1_m = std::real(sigma1_m);
+			sigma1_p = std::real(sigma1_p);
+		} // else
+			// std::cerr << "Ignoring t = " << ts[i] << std::endl;
+
+		if (std::abs(std::imag(sigma2_m)) < 1e-4) {
+			sigma2_m = std::real(sigma2_m);
+			sigma2_p = std::real(sigma2_p);
+		} // else
+			// std::cerr << "Ignoring t = " << ts[i] << std::endl;
+
+		//% Now check to see which pair pass. Only a single pair should pass, in theory.
+		//% If not, issue a warning.
+		static constexpr T my_eps = 1.0;
+
+		if (std::abs(H + J*sigma1_m + K*sigma2_m + L*sigma1_m*sigma2_m) < my_eps) {
+			sigmas1[i][sigmas1_len[i]] = sigma1_m.real();
+			sigmas2[i][sigmas1_len[i]++] = sigma2_m.real();
+		}
+		if (std::abs(H + J*sigma1_p + K*sigma2_m + L*sigma1_p*sigma2_m) < my_eps) {
+			// if (sigmas1_len[i] != 0) // !isempty(sigmas1[i])
+		  //		std::cerr << "more than one sigma1, sigma2 pair satisfies the 3rd constraint" << std::endl;
+			sigmas1[i][sigmas1_len[i]] = sigma1_p.real();
+			sigmas2[i][sigmas1_len[i]++] = sigma2_m.real();
+		}
+		if (std::abs(H + J*sigma1_p + K*sigma2_p + L*sigma1_p*sigma2_p) < my_eps) {
+			// if (sigmas1_len[i] != 0)
+      //	std::cerr << "more than one sigma1, sigma2 pair satisfies the 3rd constraint" << std::endl;
+			sigmas1[i][sigmas1_len[i]] = sigma1_p.real();
+			sigmas2[i][sigmas1_len[i]++] = sigma2_p.real();
+		}
+		if (std::abs(H + J*sigma1_m + K*sigma2_p + L*sigma1_m*sigma2_p) < my_eps) {
+			// if (sigmas1_len[i] != 0)
+			// std::cerr << "more than one sigma1, sigma2 pair satisfies the 3rd constraint" << std::endl;
+			sigmas1[i][sigmas1_len[i]] = sigma1_m.real();
+			sigmas2[i][sigmas1_len[i]++] = sigma2_p.real();
+		}
+		// if (sigmas1_len[i] == 0) // isempty(sigmas1[i])
+		// std::cerr << "no sigma1, sigma2 pair satisfies the 3rd constraint" << std::endl;
+	}
+}
+
+template<typename T>
+inline void
+invm3x3(const T (&input_m)[3][3], T (&output_m)[3][3])
```

Comment:
Why not use Eigen?

Reply:
You could easily map memory from your memory pattern to Eigen using map

Reply:
The only reason is that using Eigen would add an additional dependency on the original code for p2pt (http://github.com/rfabbri/diffgeom2pose). The way it is, it is slightly easier to copy back changes to/from the original diffgeom2pose repository.
We can use eigen no problem for OpenMVG if you prefer.

Reply:
I'm good with this. Now I have historical context. We can replace later once we have functional test with SFM pipelines.